### PR TITLE
Dev/notify monthly reports by mail

### PR DESCRIPTION
**Pull request by Pujyoooo**

### CODE REVIEW

File: app/mailers/notify.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Notify < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.notify.monthly_report_registration.subject
+  #
+  def monthly_report_registration(user_id,monthly_report_id)
```
Comment (freezer530000) [inline app/mailers/notify.rb]: **[rails_best_practices]**
- use multipart/alternative as content_type of email [:ref](http://rails-bestpractices.com/posts/2010/08/05/use-multipart-alternative-as-content_type-of-email/)

File: app/mailers/notify.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class Notify < ApplicationMailer
+
```
Comment (freezer530000) [inline app/mailers/notify.rb]: **[rubocop]**
- Extra empty line detected at class body beginning.

File: app/mailers/notify.rb (hunk 1 above)
Comment (freezer530000) [inline app/mailers/notify.rb]: **[rubocop]**
- Space missing after comma. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: app/controllers/monthly_reports_controller.rb [hunk 3]
```diff
@@ -25,6 +25,9 @@ def create
     end
 
     if @monthly_report.save
+      if @monthly_report.status == "shipped"
```
Comment (freezer530000) [inline app/controllers/monthly_reports_controller.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: app/controllers/monthly_reports_controller.rb [hunk 4]
```diff
@@ -25,6 +25,9 @@ def create
...
+        Notify.monthly_report_registration(@monthly_report.user.id,@monthly_report.id).deliver_now
```
Comment (freezer530000) [inline app/controllers/monthly_reports_controller.rb]: **[rubocop]**
- Space missing after comma. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: app/controllers/monthly_reports_controller.rb [hunk 5]
```diff
@@ -41,6 +44,9 @@ def update
     assign_relational_params(@monthly_report)
 
     if @monthly_report.save
+      if @monthly_report.status == "shipped"
```
Comment (freezer530000) [inline app/controllers/monthly_reports_controller.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: app/controllers/monthly_reports_controller.rb [hunk 6]
```diff
@@ -41,6 +44,9 @@ def update
...
+        Notify.monthly_report_registration(@monthly_report.user.id,@monthly_report.id).deliver_now
```
Comment (freezer530000) [inline app/controllers/monthly_reports_controller.rb]: **[rubocop]**
- Space missing after comma. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: spec/requests/monthly_report_spec.rb [hunk 7]
```diff
@@ -41,6 +41,7 @@
     let(:report_params) { new_report.attributes.reject { |k, _| k =~ /id\z/ || k =~ /_at/ } }
     let(:user_report) { MonthlyReport.find_by(user: user) }
 
+
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Extra blank line detected.

File: spec/requests/monthly_report_spec.rb [hunk 8]
```diff
@@ -56,6 +57,22 @@
       it { expect(user_report.present?).to eq false }
     end
 
+    context 'when shipped' do
+      let(:post_params) { { monthly_report: report_params } }
+      after(:all) do
+        ActionMailer::Base.deliveries.clear
+      end
+      subject { post monthly_reports_path, post_params }
+      it { expect{ subject }.to change{ ActionMailer::Base.deliveries.size }.by(1) }
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Space missing to the left of {.

File: spec/requests/monthly_report_spec.rb [hunk 9]
```diff
@@ -99,20 +116,40 @@
...
-      it { expect(response).to have_http_status :redirect }
-      it { expect(user_report).to be_present }
-    end
+        it { expect(response).to have_http_status :redirect }
+        it { expect(user_report).to be_present }
+        it { expect(ActionMailer::Base.deliveries.size).to eq(1) }
+      end
 
+      context 'when wip' do
+       let(:wip_report) { build(:monthly_report, :wip) }
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Use 2 (not 1) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: spec/requests/monthly_report_spec.rb [hunk 10]
```diff
@@ -99,20 +116,40 @@
...
+       let(:wip_report_params) { wip_report.attributes.reject { |k, _| k =~ /id\z/ || k =~ /_at/ } }
+       let(:patch_params) { { monthly_report: wip_report_params } }
+        before do
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Inconsistent indentation detected.

File: spec/requests/monthly_report_spec.rb [hunk 11]
```diff
@@ -99,20 +116,40 @@
...
+          patch monthly_report_path report, wip_report_params
+        end
+       it { expect(ActionMailer::Base.deliveries.size).to eq(0) }
+      end
+    end
+
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Trailing whitespace detected. [:ref](https://github.com/bbatsov/ruby-style-guide#no-trailing-whitespace)

File: spec/requests/monthly_report_spec.rb [hunk 12]
```diff
@@ -99,20 +116,40 @@
...
+       it { expect(ActionMailer::Base.deliveries.size).to eq(0) }
+      end
+    end
+  
     context 'not_found' do
       before { patch monthly_report_path report, patch_params }
       it { expect(response).to have_http_status :not_found }
       it { expect(user_report).to be_nil }
     end
+
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb]: **[rubocop]**
- Extra empty line detected at block body end.

File: spec/mailers/notify_spec.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
+require "rails_helper"
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: spec/mailers/notify_spec.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+RSpec.describe Notify, type: :mailer do
+  let!(:report) { create(:monthly_report, :with_comments, :with_tags) }
+  let(:user) { create(:user) }
+  let(:from) { "システム管理者 "+"#{Settings.mailer[:from]}" }
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)
- Surrounding space missing for operator `+`. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)
- Trailing whitespace detected. [:ref](https://github.com/bbatsov/ruby-style-guide#no-trailing-whitespace)

File: spec/mailers/notify_spec.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+RSpec.describe Notify, type: :mailer do
+  let!(:report) { create(:monthly_report, :with_comments, :with_tags) }
+  let(:user) { create(:user) }
+  let(:from) { "システム管理者 "+"#{Settings.mailer[:from]}" } 
+
+  after(:all) do
+    ActionMailer::Base.deliveries.clear
+  end
+
+  describe "monthly_report_registration" do
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: spec/mailers/notify_spec.rb [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    let(:mail) { Notify.monthly_report_registration(user.id,report.id) }
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Space missing after comma. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: spec/mailers/notify_spec.rb [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    let(:name) { user.name }
+    let(:created_at) { report.target_month.strftime('%Y年%m月') }
+    let(:title) { "#{name}が#{created_at}の月報を登録しました" }
+
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Trailing whitespace detected. [:ref](https://github.com/bbatsov/ruby-style-guide#no-trailing-whitespace)

File: spec/mailers/notify_spec.rb [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    ActionMailer::Base.deliveries.clear
+  end
+
+  describe "monthly_report_registration" do
+    let(:mail) { Notify.monthly_report_registration(user.id,report.id) }
+    let(:name) { user.name }
+    let(:created_at) { report.target_month.strftime('%Y年%m月') }
+    let(:title) { "#{name}が#{created_at}の月報を登録しました" }
+    
+    it "deliver_now" do
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: spec/mailers/notify_spec.rb [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+      expect { mail.deliver_now }.to change{ ActionMailer::Base.deliveries.count }.from(0).to(1)
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Space missing to the left of {.

File: spec/mailers/notify_spec.rb [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    it "renders the headers" do
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: spec/mailers/notify_spec.rb [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+      expect(mail.subject).to eq(title)
+      expect(mail.to).to include(user.email)
+      expect(mail.from).to eq(from)
+    end
+
+    it "renders the body" do
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Prefer single-quoted strings when you don't need string interpolation or special symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)

File: spec/mailers/notify_spec.rb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+      expect(mail.body.encoded).to match(report.project_summary)
+      expect(mail.body.encoded).to match(user.name)
+    end
+  end
+
```
Comment (freezer530000) [inline spec/mailers/notify_spec.rb]: **[rubocop]**
- Extra empty line detected at block body end.

File: spec/mailers/previews/notify_preview.rb [hunk 23]
```diff
@@ -0,0 +1,9 @@
+# Preview all emails at http://localhost:3000/rails/mailers/notify
+class NotifyPreview < ActionMailer::Preview
+
```
Comment (freezer530000) [inline spec/mailers/previews/notify_preview.rb]: **[rubocop]**
- Extra empty line detected at class body beginning.

File: spec/mailers/previews/notify_preview.rb [hunk 24]
```diff
@@ -0,0 +1,9 @@
...
+  # Preview this email at http://localhost:3000/rails/mailers/notify/monthly_report_registration
+  def monthly_report_registration
+    Notify.monthly_report_registration
+  end
+
```
Comment (freezer530000) [inline spec/mailers/previews/notify_preview.rb]: **[rubocop]**
- Extra empty line detected at class body end.

**Comment by freezer530000** [comment]
Click here if you want to create rubocop's auto-correct Pull Request 
[](https://sideci.com/github_repositories/42654368/auto_correct/rubocop/pull_request/240/branch/dev%2Fnotify_monthly_reports_by_mail)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/mailers/notify.rb [hunk 25]
```diff
@@ -0,0 +1,13 @@
+class Notify < ApplicationMailer
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.notify.monthly_report_registration.subject
+  #
+  def monthly_report_registration(user_id, monthly_report_id)
```
Comment (freezer530000) [inline app/mailers/notify.rb:7]: **[rails_best_practices]**
- use multipart/alternative as content_type of email [:ref](http://rails-bestpractices.com/posts/2010/08/05/use-multipart-alternative-as-content_type-of-email/)

File: spec/requests/monthly_report_spec.rb [hunk 26]
```diff
@@ -87,6 +87,7 @@
         it { expect(response).to have_http_status :redirect }
         it { expect(user_report.present?).to eq true }
       end
+
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb:90]: **[rubocop]**
- Extra empty line detected at block body end.

File: spec/mailers/previews/notify_preview.rb [hunk 27]
```diff
@@ -0,0 +1,9 @@
+# Preview all emails at http://localhost:3000/rails/mailers/notify
```
Comment (Pujyoooo) [inline spec/mailers/previews/notify_preview.rb:1]: 消す

File: spec/fixtures/notify/monthly_report_registration [hunk 28]
```diff
@@ -0,0 +1,3 @@
+Notify#monthly_report_registration
```
Comment (Pujyoooo) [inline spec/fixtures/notify/monthly_report_registration:1]: 消す

File: app/mailers/notify.rb [hunk 29]
```diff
@@ -5,9 +5,10 @@ class Notify < ApplicationMailer
   #
   #   en.notify.monthly_report_registration.subject
   #
-  def monthly_report_registration
-    @greeting = "Hi"
-
-    mail to: Settings.mailer[:mail_address]
+  def monthly_report_registration(user_id,monthly_report_id)
+    @report = MonthlyReport.find(monthly_report_id)
+    @user = User.find(user_id)
```
Comment (Pujyoooo) [inline app/mailers/notify.rb:9]: current_userで渡す

File: app/views/notify/monthly_report_registration.html.slim [hunk 30]
```diff
@@ -0,0 +1,23 @@
+p 今月の月報が登録されました。
```
Comment (Pujyoooo) [inline app/views/notify/monthly_report_registration.html.slim:1]: 今月とは限らないので〜月の月報にする

File: config/settings/development.yml [hunk 31]
```diff
@@ -1,3 +1,4 @@
 mailer:
   mail_address: 'hoge@example.com'
```
Comment (Pujyoooo) [inline config/settings/development.yml]: mail_addressは消す

File: spec/mailers/notify_spec.rb [hunk 32]
```diff
@@ -1,17 +1,32 @@
 require "rails_helper"
-
```
Comment (Pujyoooo) [inline spec/mailers/notify_spec.rb]: 改行する

File: spec/mailers/notify_spec.rb [hunk 33]
```diff
@@ -1,17 +1,32 @@
...
     it "renders the headers" do
-      expect(mail.subject).to eq("Monthly report registration")
-      expect(mail.to).to eq(["to@example.org"])
-      expect(mail.from).to eq(["from@example.com"])
+      expect(mail.subject).to eq(title)
+      expect(mail.to).to include(user.email)
+      expect(mail.from[0]).to eq(from)
     end
 
     it "renders the body" do
```
Comment (Pujyoooo) [inline spec/mailers/notify_spec.rb]: it の中で複数expectしてるので分割する

File: app/controllers/monthly_reports_controller.rb [hunk 34]
```diff
@@ -110,4 +111,10 @@ def search_params
 
     params.require(:q).permit(search_conditions)
   end
+
+  def monthly_report_notify
+    user_id = @monthly_report.user.id
+    report_id = @monthly_report.id
+    Notify.monthly_report_registration(user_id, report_id).deliver_now if @monthly_report.status == 'shipped'
```
Comment (Pujyoooo) [inline app/controllers/monthly_reports_controller.rb:118]: 初めてshipped_atに値が入ってきた時のみ送るようにする。

File: app/views/notify/monthly_report_registration.html.slim [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+
+p 登録日時: #{@report.created_at}
+
+p 氏名: #{@user.name}
+
+p 今月の業務報告
+p = "-------------------------------------------------------------"
+p 【プロジェクト概要】
+= simple_format(h(@report.project_summary))
+p 【使用した技術】
```
Comment (Pujyoooo) [inline app/views/notify/monthly_report_registration.html.slim:11]: プロジェクト概要以下は削除して
レポートへのリンクを貼る。

File: app/views/notify/monthly_report_registration.html.slim [hunk 36]
```diff
@@ -0,0 +1,23 @@
+p 今月の月報が登録されました。
+
+p 登録日時: #{@report.created_at}
+
+p 氏名: #{@user.name}
+
+p 今月の業務報告
```
Comment (Pujyoooo) [inline app/views/notify/monthly_report_registration.html.slim:7]: 今月とは限らないので〜月の月報にする

File: config/settings.yml [hunk 37]
```diff
@@ -0,0 +1,10 @@
+mailer:
```
Comment (Pujyoooo) [inline config/settings.yml]: 環境変数を参照するように変更

File: spec/mailers/notify_spec.rb [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+    let(:name) { user.name }
+    let(:created_at) { report.target_month.strftime('%Y年%m月') }
+    let(:title) { "#{name}が#{created_at}の月報を登録しました" }
+
+    it 'deliver_now' do
+      expect { mail.deliver_now }.to change { ActionMailer::Base.deliveries.count }.from(0).to(1)
+    end
+
+    it 'renders the headers' do
+      expect(mail.subject).to eq(title)
```
Comment (Pujyoooo) [inline spec/mailers/notify_spec.rb:22]: itの中に複数のexpectがあるので分割する

File: spec/mailers/notify_spec.rb [hunk 39]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+RSpec.describe Notify, type: :mailer do
```
Comment (Pujyoooo) [inline spec/mailers/notify_spec.rb:2]: 改行いれる

File: spec/requests/monthly_report_spec.rb [hunk 40]
```diff
@@ -176,26 +197,51 @@
...
     let(:process_params) { [build(:monthly_working_process).process] }
     let(:user_report) { MonthlyReport.find_by(report_params, user: report.user) }
     let(:patch_params) do
       {
         monthly_report: report_params.merge(monthly_report_tags: tag_params),
         working_process: process_params,
-        wip: true,
       }
     end
-
```
Comment (Pujyoooo) [inline spec/requests/monthly_report_spec.rb]: 消しがち。
消しちゃあかん

File: spec/requests/monthly_report_spec.rb [hunk 41]
```diff
@@ -176,26 +197,51 @@
...
       end
 
-      it { expect(response).to have_http_status :redirect }
-      it { expect(user_report).to be_present }
+      describe 'notify mail' do
+        after(:each) do
+          ActionMailer::Base.deliveries.clear
+        end
+
+        context 'when shipped' do
```
Comment (Pujyoooo) [inline spec/requests/monthly_report_spec.rb:228]: 事前にshippedのレポートのテストケースと
shippedに変わったレポートのテストケースが必要

File: db/schema.rb [hunk 42]
```diff
@@ -31,45 +45,33 @@
...
+    t.string   "help_type"
+    t.string   "target"
+    t.text     "body"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
   end
 
+  add_index "help_texts", ["category", "help_type"], name: "index_help_texts_on_category_and_help_type", using: :btree
+
+  create_table "monthly_report_comments", force: :cascade do |t|
```
Comment (freezer530000) [inline db/schema.rb:67]: **[rails_best_practices]**
- always add db index (monthly_report_comments => [user_id]) [:ref](http://rails-bestpractices.com/posts/2010/07/24/always-add-db-index/)
- always add db index (monthly_report_comments => [monthly_report_id]) [:ref](http://rails-bestpractices.com/posts/2010/07/24/always-add-db-index/)

File: config/routes.rb [hunk 43]
```diff
@@ -1,6 +1,38 @@
...
+      end
+    end
+  end
+
   ActiveAdmin.routes(self)
+
+  resources :groups, only: [:index]
+  resources :monthly_report_comments, only: [:create, :edit, :update, :destroy]
+
+  resources :sample, only: [:index] do
```
Comment (freezer530000) [inline config/routes.rb:17]: **[rails_best_practices]**
- overuse route customizations (customize_count > 3) [:ref](http://rails-bestpractices.com/posts/2010/07/22/overuse-route-customizations/)

File: config/routes.rb [hunk 44]
```diff
@@ -1,6 +1,38 @@
 Rails.application.routes.draw do
-  devise_for :admin_users, ActiveAdmin::Devise.config
+
+  devise_for :users, skip: [:sessions]
+  devise_scope :user do
+    resources :sessions, only: [:new, :create] do
```
Comment (freezer530000) [inline config/routes.rb:5]: **[rails_best_practices]**
- restrict auto-generated routes sessions (only: []) [:ref](http://rails-bestpractices.com/posts/2011/08/19/restrict-auto-generated-routes/)

File: spec/requests/monthly_report_spec.rb [hunk 45]
```diff
@@ -0,0 +1,253 @@
...
+          {
+            monthly_report: report_params.merge(monthly_report_tags: tag_params),
+            working_process: process_params,
+          }
+        end
+
+        it { expect(response).to have_http_status :redirect }
+        it { expect(user_report.present?).to eq true }
+      end
+
```
Comment (freezer530000) [inline spec/requests/monthly_report_spec.rb:90]: **[rubocop]**
- Extra empty line detected at block body end.

**Pull request closed without merging**